### DISCUSSION TITLE
Dependent select morph issue

### DISCUSSION THREAD

**Discussion by austincarpenter** [livewire/livewire#6882] [7. Bugs (archived)]
### Livewire version

v3.0.5

### Which PHP version are you using?

PHP 8.2

### Repository URL

https://github.com/austincarpenter/select-bug-livewire-v3

### Steps To Reproduce

With Livewire 2, when a dependent select's options were changed to no longer include the currently selected option (in terms of its value), it would default back to having selected a blank option.

https://github.com/livewire/livewire/assets/3736774/cfa1cd1f-9b0f-4828-82a6-35a973b6a2ef

With Livewire 3, once the options are changed, the select will show the nth option available (where n is the selected option), despite its value being different.

https://github.com/livewire/livewire/assets/3736774/ec6cb341-519a-48ab-a685-46662130a9c6

Latest commit on demo repo is v3, previous is v2.

Not sure if this behaviour is intended or not with the new version.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by austincarpenter** [comment]
Fix: I can see now that of course `wire:key` is needed on each of the options iterated through `@foreach`, although clearly the behaviour differs between versions. Will PR an update to the forms docs.

**Reply by austincarpenter** [reply]
#6883

**Comment by wizzymore** [comment]
I think a better solution would have been to wire:key the dependant select based on the category, not all of the childs.

**Reply by austincarpenter** [reply]
Not sure what you mean here as `wire:key` is meant to help the differ when enumerating through an array.

https://livewire.laravel.com/docs/components#adding-wirekey-to-foreach-loops

**Reply by wizzymore** [reply]
Yes, it is, but the array that you have is changing when based on the first select which will always contain a "single" value ( in your example ). So you can do a smarter `wire:key` and have it check the change only in one place instead of in all of the options.

`wire:key` is very good for the cases when you will need to update a specific row, a very targeted update, here you target is not really the "Options" themselves ( again, based on your example ) it's the select. This is because all the values inside the select will change instead of a targeted few.

Let's say you have a select with all the countries and another select that is dependant on the first one that contains all the states in the country. You know from the start that they won't repeat when the country is changing so the better approach is to wire:key the country value not each individual state. This is 1 check vs `n` checks. Also this helps you because the dependant select will be re-rendered completely so the state will be "reset", this means you will not have conflicts between country changes (in this example)

**Reply by austincarpenter** [reply]
Apologies I missed this one. Thanks for the tip. Can't remember if I had tried it on the parent. https://laravel-livewire.com/docs/2.x/troubleshooting#dom-diffing-cures recommends the children receive a `wire:key` within a `@foreach` but if it works on the parent (as it does now following #6889), fantastic.
